### PR TITLE
cache umt5 bias computation

### LINKED ISSUE #26144: UMT5 incredibly slow in generating
### System Info

- `transformers` version: 4.33.1
- Platform: Linux-5.14.0-284.25.1.el9_2.x86_64-x86_64-with-glibc2.34
- Python version: 3.11.4
- Huggingface_hub version: 0.16.4
- Safetensors version: 0.3.1
- Accelerate version: 0.22.0
- Accelerate config:    not found
- PyTorch version (GPU?): 2.0.1+cu117 (True)

### Who can help?

@ArthurZucker and @younesbelkada and @gante

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```python
import time

from transformers import AutoTokenizer, AutoModelForSeq2SeqLM, GenerationConfig


if __name__ == "__main__":
    timings = {}

    for model_name in ("facebook/mbart-large-50-many-to-one-mmt", "google/umt5-small"):
        model = AutoModelForSeq2SeqLM.from_pretrained(model_name, device_map={"": "cuda"})
        print(model_name, model.num_parameters())
        # google/umt5-small                        306601984
        # facebook/mbart-large-50-many-to-one-mmt 1122990080
        tokenizer = AutoTokenizer.from_pretrained(model_name)
        gen_config = GenerationConfig.from_pretrained(
            model_name,
            max_new_tokens=200,
            max_length=None,
            num_beams=1,
        )
        text = "I would really like to eat some cookies now."
        if "t5" in model_name:
            text = f"translate English to Dutch: {text}"

        encoded = tokenizer(text, return_tensors="pt")
        encoded = {k: v.to(model.device) for k, v in encoded.items()}
        start_time = time.perf_counter_ns()
        for _ in range(100):
            _ = model.generate(**encoded, generation_config=gen_config)

        timings[model_name] = time.perf_counter_ns() - start_time

    for model_name, timings in timings.items():
        print(f"Generation duration for {model_name.split('/')[1]}:\t{timings}")
        # Generation duration for mbart-large-50-many-to-one-mmt:  22413427363
        # Generation duration for umt5-small:                     207906791077
```

So despite UMT5-small having only about **27%** the number of parameters of the MBART-large model it is **9-10x** slower!

(I also tried with a gc.collect() after each generation.)

### Expected behavior

Faster inference/generation speed. Training is fine so I assume caching of past states is not (correctly) implemented but I might be wrong. This PR on adding caching to T5 by @patrickvonplaten might be related: https://github.com/huggingface/transformers/pull/3682

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Should fix #26144, by caching the relative bias and only returning the one that is required for the current computation.

### CODE REVIEW

**Comment by BramVanroy** [comment]
This is great @ArthurZucker! My new results are:

```
Comparing to facebook/mbart-large-50-many-to-one-mmt:
-----------------------------------
google/umt5-small:       22.08% (relative speed); 50.19% (relative size)
google/mt5-small:        12.57% (relative speed); 49.14% (relative size)
t5-small:        32.21% (relative speed); 9.90% (relative size)
```

That's like a 45x improvement?! :o

I know I'm probably _too_ excited but any idea when this can be merged? Or is https://github.com/huggingface/transformers/issues/23145 blocking this?

**Comment by ArthurZucker** [comment]
The linked issue is blocking + for now the output is garbage, working on this 😉

**Comment by BramVanroy** [comment]
Ah great, I'll have some patience and jeep an eye out to the progress here.

File: src/transformers/models/umt5/modeling_umt5.py [hunk 1]
```diff
@@ -223,23 +199,61 @@ def _relative_position_bucket(self, relative_position):
...
+        memory_position = torch.arange(key_length, dtype=torch.long, device = self.weight.device)[None, :]
         relative_position = memory_position - context_position  # shape (query_length, key_length)
         relative_position_bucket = self._relative_position_bucket(relative_position)
-        values = self.relative_attention_bias(relative_position_bucket)  # shape (query_length, key_length, num_heads)
-        values = values.permute([2, 0, 1]).unsqueeze(0)  # shape (1, num_heads, query_length, key_length)
-        return values
+        values = super().forward(relative_position_bucket).to(self.weight.device) # shape (query_length, key_length, num_heads)
+        # shape (1, num_heads, query_length, key_length)
+        cached_bias = values.permute([2, 0, 1]).unsqueeze(0)
+        self.register_buffer("cached_bias", cached_bias, persistent=False)
```
Comment (SunMarc) [inline src/transformers/models/umt5/modeling_umt5.py:217]: When we use `device_map`, the non persistent buffer is on the meta device (comment above). Moreover, as the value is not present in the `state dict`,  `cached_bias` will stay on the meta device when we are loading the weights . Then, when we try to dispatch the model to cuda, the following error will appear : `Cannot copy out of meta tensor; no data!`

**Review by SunMarc — COMMENTED** [review]
Left a few comments to explain why the model doesn't work with `device_map`

**Comment by ArthurZucker** [comment]
Thanks! Did not have time to improve this yet :(

**Comment by BramVanroy** [comment]
> Thanks! Did not have time to improve this yet :(

No worries! I'll just keep bumping this until you find some time! (Or if you want to just close this that's only fine ofc)

**Comment by ArthurZucker** [comment]
Closing as I won't have time to tackle this! 🤗

**Pull request closed without merging**